### DISCUSSION TITLE
Extremely noisy defaults (recommended rules)

### DISCUSSION THREAD

**Discussion by harjotgill** [biomejs/biome#3129] [General]
We recently rolled out Biome to our large customer base at CodeRabbit. We were under the assumption that, just like Ruff, Biome would have recommended rules that everyone agreed upon. However, the roll-out has been a disaster, and we ended up with a lot of cancellations and complaints on all our channels. Users were pulling their hair and screaming at the AI bot to spare them from noisy suggestions. In particular, `useImportType`, `useExhaustiveDependencies ` and `noForEach` decimated the user experience. 

For a linter, the defaults shouldn't be more than 10% of the total rules. In case of Biome, 90% rules are enabled as defaults - something is clearly off here. 

I suggest that the maintainers take a deep look at the rule list and disable any rule that comes across as an opinion or has caveats. Defaults should only be those rules that everyone can agree upon. Ruff has done a great job with defaults, and that is one of the reasons it has good traction.

**Comment by ematipico** [comment]
Thank you for your feedback! It's a pity that the rollout didn't go as expected. 

> We recently rolled out Biome to our large customer base at CodeRabbit. We were under the assumption that, just like Ruff, Biome would have recommended rules that everyone agreed upon. However, the roll-out has been a disaster, and we ended up with a lot of cancellations and complaints on all our channels. Users were pulling their hair and screaming at the AI bot to spare them from noisy suggestions. In particular, `useImportType`, `useExhaustiveDependencies ` and `noForEach` decimated the user experience.
> 

I admit that maybe a few rules don't need to be recommended. However, in this particular example, I would argue that `useExhaustiveDependencies` should be recommended because the React core team recommends it. 

> For a linter, the defaults shouldn't be more than 10% of the total rules.

Why it should be 10%? Are there articles or talks that you can share that explain the reason behind this?

> In case of Biome, 90% rules are enabled as defaults - something is clearly off here. 

I admit that we have a lot of recommended rules, however this information needs some **important context**. Biome lints **three languages**: JavaScript, JSX and TypeScript. These languages come with their quirks, patterns and side-effects. Plus, we recently rolled out CSS support, which is a very relaxed language, which requires *more* rules in order to build correct CSS. At the moment we have around 15 nursery rules for CSS, if we had to follow *your 10% suggestion*, we would need to recommend 2 rules, which doesn't work.

Plus, considering that Biome is very recent, and we already have **tons** of recommended rules in the ecosystem (eslint + plugins + other linters), have you considered that maybe we decided to write first those rules that are generally used by the majority of projects out there (and recommended)? A11y rules are a clear example. They are many, and I don't see any strong reason to not recommend them.

> Defaults should only be those rules that everyone can agree upon.

Who is "everyone" in this case? :) Biome a **young** project with a **small** team and a **small** community 😄 so if the team and the small number of contributors that follow the project feel good about some defaults, we ship it. We have a relaxed process at the moment. Do you have any particular suggestions? Any particular idea in mind?

**Reply by lostpebble** [reply]
I think an issue I've found lately is that there are loads of new recommendations- but they are all **errors**. So in some ways, these are more than just "recommendations" now- it feels forced, adding biome to a well established project is sure to have LOADS of errors everywhere now, and I end up converting them to "warn" or otherwise "off" (if its not something I will be able to fix soon and the warning will be annoying to look at all the time).

A good way to get around this might be to by default make new rules "warn" instead? And only make the most serious of things "error" by default.

**Reply by ematipico** [reply]
How would you define the "most serious ones"? What kind of metrics would you use? It would be great to try to craft a framework that would help us to point us in the right direction

**Reply by lostpebble** [reply]
Probably everything under `style`, `a11y`, `complexity` and `correctness` could be warnings instead of errors. But I agree with you, its hard to define what is a **grievous error** and what could be considered a strong recommendation (still should be a warning).

**Reply by Conaclos** [reply]
I previously proposed for Biome 2.0 to associate a de fault severity level at the group level: I proposed `error` for correctness and `warn` for all other groups.

**Reply by Conaclos** [reply]
Personally, I am in favour of having a stronger default than eslint. However, we can discuss - as we are doing here - about how to reduce possible noise: for example by reducing false positives, deprecating or removing from recommended the rules with low value, ...

**Reply by harjotgill** [reply]
If it helps, this is what we are setting as defaults. Note that CodeRabbit has 2 AI profiles - "Chill" and "Assertive". We are setting the defaults based on the profile.

```
	const biomeConfig = {
		$schema: "https://biomejs.dev/schemas/1.7.3/schema.json",
		organizeImports: {
			enabled: false,
		},
		formatter: {
			enabled: false,
		},
		linter: {
			enabled: true,
			rules: {
				all: false,
				recommended: true,
				correctness: {
					useExhaustiveDependencies: "off",
				},
				style: {
					recommended: isAssertive,
					noArguments: "error",
					noCommaOperator: "error",
					noUselessElse: "error",
					useAsConstAssertion: "off",
					useBlockStatements: "off",
					useConsistentArrayType: "off",
					useDefaultParameterLast: "error",
					useEnumInitializers: "off",
					useExponentiationOperator: "error",
					useExportType: "off",
					useFragmentSyntax: "off",
					useImportType: "off",
					useLiteralEnumMembers: "error",
					useShorthandArrayType: "off",
					noUnusedTemplateLiteral: "off",
				},
				complexity: {
					noForEach: "off",
					noExcessiveCognitiveComplexity: isAssertive ? "error" : "off",
					useLiteralKeys: "off",
				},
				suspicious: {
					noArrayIndexKey: isAssertive ? "error" : "off",
					noExplicitAny: isAssertive ? "error" : "off",
					noImplicitAnyLet: isAssertive ? "error" : "off",
					noDoubleEquals: isAssertive ? "error" : "off",
				},
				nursery: {
					all: false,
				},
			},
		},
		css: {
			linter: {
				enabled: true,
			},
		},
	}
```

**Comment by fregante** [comment]
> the defaults shouldn't be more than 10% of the total rules

> Defaults should only be those rules that everyone can agree upon

Disagree on both points. Such number is completely arbitrary. Everyone has opinions, defaults should point users to improve their codebases. You might not like a rule, but if it's backed by an argument for its utility, then you should probably use it.

The only thing I can agree with is that enabling it on an unlinted repository is going to be tough… but that's just a consequence of adding a bunch of rules at once. I don't see an issue with Biome itself.

Prettier and Standard became popular because they prompt you to use a bunch of style/rules that have been discussed over years as best practices. If you disagree, don't use those projects.

This is right there on the homepage:

> Simple
> Zero configuration needed to get started. **Extensive options available for when you need them.**
